### DISCUSSION TITLE
eleventyComputed inside global data files

### DISCUSSION THREAD

**Discussion by nachtfunke** [11ty/eleventy#2397] [Uncategorized]
Hello! I am trying to understand something about eleventyComputed that doesn't make much sense to me.

I have a directory `11tydata.js` file, that had a structure like this:

```javascript
eleventyComputed:{
    api: {
        url: process.env.API_URL,
        base: data => `${data.api.url}/api/v1`,
        methods: {
            getCustomerData: data => `${data.api.base}/bla/getbyid`,
            create: data => `${data.api.base}/bla/create`
        }
    }
}
```

I then turn this into a json file with some filters. This worked perfectly fine. But then I realised that it would make more sense to move this into a global data file, `api.js`:

```javascript
module.exports = {
    url: process.env.API_URL,
    base: data => `${data.url}/api/v1`,
    methods: {
        getCustomerData: data => `${data.base}/bla/getbyid`,
        create: data => `${data.base}/bla/create`
    }
}
```
But to my surprise, the generated file only contained `url` with the correct data, and an empty `methods: {}`. I tried accessing the data with `api.methods.create` and it printed it out fine.

What is going on here? I thought that within global data files, the `eleventyComputed` should be omitted, according to #1104.

**Comment by pdehaan** [comment]
@nachtfunke Sorry for the late reply...
I think you'll still need the `eleventyComputed`, but I wasn't able to get that working yet in your global data file.

In the meantime, since you aren't using any computed data from outside the data file itself, this might be a workaround:

```js
const url = process.env.API_URL;
const base = new URL("/api/v1", url).href;

module.exports = {
  url,
  base,
  methods: {
    getCustomerData: `${base}/bla/getbyid`,
    create: `${base}/bla/create`
  }
};
```

**Reply by pdehaan** [reply]
Alright, not sure how I feel about it, but I got it working with a global _data/eleventyComputed.js file:

```js
module.exports = {
  api: {
    url: process.env.API_URL,
    base: (data) => `${data.api.url}/api/v1`,
    methods: {
      getCustomerData: (data) => `${data.api.base}/bla/getbyid`,
      create: (data) => `${data.api.base}/bla/create`,
    },
  }
};
```

```liquid
---
# index.liquid
---
url={{ api.url }}
base={{ api.base }}
methods.getCustomerData={{ api.methods.getCustomerData }}
methods.create={{ api.methods.create }}
```

### OUTPUT
```html
url=http://lorem.ipsum
base=http://lorem.ipsum/api/v1
methods.getCustomerData=http://lorem.ipsum/api/v1/bla/getbyid
methods.create=http://lorem.ipsum/api/v1/bla/create
```

---

**UPDATE:** Actually, looking at this tab again in the morning, this might be a slightly cleaner _data/eleventyComputed.js solution:

```js
// _data/eleventyComputed.js
module.exports = {
  api(data) {
    const url = process.env.API_URL;
    const base = `${url}/api/v1`;
    return {
      url,
      base,
      methods: {
        getCustomerData: `${base}/bla/getbyid`,
        create: `${base}/bla/create`
      }
    }
  }
};
```

I'm personally not a fan of global eleventyComputed.js files, and would either go with a directory data file or probably something like my initial workaround in https://github.com/11ty/eleventy/discussions/2397#discussioncomment-2800981 since the data is pretty basic/contained.
